src/shared/projectConfiguration.ts

```diff
@@ -0,0 +1,67 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as crypto from 'crypto';
+import { machineIdSync } from 'node-machine-id';
+import { PlatformInformation } from './platform';
+import { ITelemetryReporter, getTelemetryProps } from './telemetryReporter';
+import { DotnetInfo } from './utils/dotnetInfo';
+
+export interface ProjectConfigurationMessage {
+    ProjectId: string;
+    SessionId: string;
+    OutputKind: number;
+    ProjectCapabilities: string[];
+    TargetFrameworks: string[];
+    References: string[];
+    FileExtensions: string[];
+    FileCounts: number[];
+    SdkStyleProject: boolean;
+}
+
+export function reportProjectConfigurationEvent(
+    reporter: ITelemetryReporter,
+    projectConfig: ProjectConfigurationMessage,
+    platformInfo: PlatformInformation,
+    dotnetInfo: DotnetInfo | undefined,
+    solutionPath?: string,
+    useModernNet?: boolean
+) {
+    let solutionId = '';
+    if (solutionPath) {
+        solutionId = createSolutionId(solutionPath);
+    }
+
+    const telemetryProps = getTelemetryProps(platformInfo);
+    telemetryProps['SolutionId'] = solutionId;
+    telemetryProps['ProjectId'] = projectConfig.ProjectId;
+    telemetryProps['SessionId'] = projectConfig.SessionId;
+    telemetryProps['OutputType'] = projectConfig.OutputKind?.toString() ?? '';
+    telemetryProps['ProjectCapabilities'] = projectConfig.ProjectCapabilities?.join(' ') ?? '';
+    telemetryProps['TargetFrameworks'] = projectConfig.TargetFrameworks.join('|');
+    telemetryProps['References'] = projectConfig.References.join('|');
+    telemetryProps['FileExtensions'] = projectConfig.FileExtensions.join('|');
+    telemetryProps['FileCounts'] = projectConfig.FileCounts?.join('|') ?? '';
+    telemetryProps['NetSdkVersion'] = dotnetInfo?.Version ?? '';
+    telemetryProps['sdkStyleProject'] = projectConfig.SdkStyleProject.toString();
+
+    if (useModernNet) {
+        telemetryProps['useModernNet'] = useModernNet.toString();
+    }
+
+    reporter.sendTelemetryEvent('ProjectConfiguration', telemetryProps);
```

Comment:
💡 It would be good to update `reporter` to throw unconditionally when telemetry is disabled so these wasted CPU cycles are harder to introduce

Reply:
I'd imagine that'll mean we just ship a bunch of bugs every time we aren't testing that way. But maybe have sendTelemetryEvent take a lambda which returns the item, and we skip that if we aren't using it?

Reply:
I will add something here in a followup - needs some work and it is an existing issue with O#.  Potentially we could make the roslyn lsp one throw (or log an error) and skip in the O# case.